### PR TITLE
Fallback path support packages

### PR DESCRIPTION
**Pull request by om26er**
Improvement to https://github.com/crossbario/crossbar/pull/1408, now supporting reading fallback from python packages as well.

closes https://github.com/crossbario/crossbar/issues/1712

### CODE REVIEW

File: crossbar/webservice/static.py [hunk 1]
```diff
@@ -155,8 +155,8 @@ def create(transport, path, config):
         #
         fallback = static_options.get('default_file')
         if fallback:
-            resource.childNotFound = ResourceFallback(path, config)
+            resource.childNotFound = ResourceFallback(os.path.join(static_dir, fallback))
```
Comment (oberstet) [inline crossbar/webservice/static.py]: we can remove `ResourceFallback` altogether and just use `File` here directly

File: crossbar/webservice/static.py [hunk 2]
```diff
@@ -131,7 +131,7 @@ def create(transport, path, config):
 
             raise ApplicationError("crossbar.error.invalid_configuration", "missing web spec")
 
-        static_dir = static_dir.encode('ascii', 'ignore')  # http://stackoverflow.com/a/20433918/884770
+        static_dir_enc = static_dir.encode('ascii', 'ignore')  # http://stackoverflow.com/a/20433918/884770
```
Comment (oberstet) [inline crossbar/webservice/static.py]: why do we need to change the variable name?

File: crossbar/webservice/base.py [hunk 3]
```diff
@@ -62,11 +62,8 @@ class ResourceFallback(File):
     """
     Handle requests for non-existent URL's
     """
-    def __init__(self, path, config, **kwargs):
+    def __init__(self, path, **kwargs):
```
Comment (oberstet) [inline crossbar/webservice/base.py]: we can completely remove `ResourceFallback` (see also the other comment below)

**Comment by oberstet** [comment]
also: ideally, we had an issue file for this, a changelog entry refering the issue/PR and adjustment to docs. (I skip a CI test .. for this PR - but I think the changelog/docs would be quick to add ..)

File: crossbar/webservice/base.py (hunk 3 above)
Comment (om26er) [reply crossbar/webservice/base.py]: done

File: crossbar/webservice/static.py (hunk 1 above)
Comment (om26er) [reply crossbar/webservice/static.py]: done

File: crossbar/webservice/static.py [hunk 4]
```diff
@@ -155,8 +155,8 @@ def create(transport, path, config):
         #
         fallback = static_options.get('default_file')
         if fallback:
-            resource.childNotFound = ResourceFallback(path, config)
+            resource.childNotFound = File(os.path.join(static_dir, fallback))
         else:
-            resource.childNotFound = Resource404(transport.templates, static_dir)
+            resource.childNotFound = Resource404(transport.templates, static_dir_enc)
```
Comment (oberstet) [inline crossbar/webservice/static.py]: needs to be "static_dir"

File: crossbar/webservice/static.py [hunk 5]
```diff
@@ -143,7 +143,7 @@ def create(transport, path, config):
         cache_timeout = static_options.get('cache_timeout', DEFAULT_CACHE_TIMEOUT)
         allow_cross_origin = static_options.get('allow_cross_origin', True)
 
-        resource = static_resource_class(static_dir, cache_timeout=cache_timeout, allow_cross_origin=allow_cross_origin)
+        resource = static_resource_class(static_dir_enc, cache_timeout=cache_timeout, allow_cross_origin=allow_cross_origin)
```
Comment (oberstet) [inline crossbar/webservice/static.py]: needs to be "static_dir"

File: crossbar/webservice/static.py (hunk 2 above)
Comment (oberstet) [reply crossbar/webservice/static.py]: needs to be "static_dir"

Comment (om26er) [reply crossbar/webservice/static.py]: We use the existing `static_dir` variable in the newly added code below, It was being encoded and overwritten previously. As `os.path.join()` expects both paths to be string.

The alternate could be to just avoid the renaming and call `static_dir.decode('ascii')` where needed. Your call!

**Comment by om26er** [comment]
> also: ideally, we had an issue file for this, a changelog entry refering the issue/PR and adjustment to docs. (I skip a CI test .. for this PR - but I think the changelog/docs would be quick to add ..)

Filed https://github.com/crossbario/crossbar/issues/1712, added a change log entry and also created another issue to track a CI test for this feature. Only need to update the docs, though I wonder what would be the "right" place for that...

**Comment by oberstet** [comment]
perfect, ready to merge!

**Pull request merged**